Add options for rake routes task

davydovanton opened the pull request:
It's simple realization for #18902. 
`-g` option returns the urls `name`, `verb`, `path` and `reqs` field that match the pattern.
`-c` option returns the urls for specific controller.

For this I use `OptionParser` from ruby standart lib. I do not like to use `--` symbol, but it's necessary for rake. Also I have a question: May be we replace `CONTROLLER` option to key option too?

davydovanton committed:
Add options for rake routes task

Add two options: `-c` and `-g`.
`-g` option returns the urls name, verb and path fields that match the pattern.
`-c` option returns the urls for specific controller.

rafaelfranca commented on guides/source/routing.md:
````diff
@@ -1115,6 +1115,11 @@ You may restrict the listing to the routes that map to a particular controller s
 
+Also you may grep your routers with `-g` options. This option returns the url `name`, `verb`, `path` and `reqs` field that match the pattern:
+```bash
+$ bin/rake routes -- -g users
````
why `--` is required?

davydovanton replied:
`--` split rake tasks and key options

rafaelfranca replied:
Can't we not use the `--`?

davydovanton commented:
ops, i need fix tests, sorry :sweat:

davydovanton commented:
@rafaelfranca i fixed `--`. Now you can use this syntax:
```
$ rake routes -g pattern
```

davydovanton commented:
@rafaelfranca @dhh what do you think if I replace `CONTROLLER` options to key option too? Now this option is set by ENV variable.

kaspth commented on actionpack/lib/action_dispatch/routing/inspector.rb:
```diff
@@ -63,10 +63,11 @@ def initialize(routes)
 
-      def format(formatter, filter = nil)
-        routes_to_display = filter_routes(filter)
+      def format(formatter, **options)
```
Why not `format(formatter, filter:, pattern:)`?

davydovanton replied:
I know that's YAGNI but I want to add a few more options, for example `--model` and `--exclude`

kaspth commented on actionpack/lib/action_dispatch/routing/inspector.rb:
```diff
@@ -86,6 +87,10 @@ def format(formatter, filter = nil)
       private
 
+      def grep_routes(routes, pattern)
+        routes.select { |r| /#{pattern}/ =~ r.values.join(' ') }
```
Change the order on the arguments being matched, so pattern is after values.

kaspth commented on guides/source/routing.md:
````diff
@@ -1115,6 +1115,11 @@ You may restrict the listing to the routes that map to a particular controller s
 $ CONTROLLER=users bin/rake routes
 ```
 
+Also you may grep your routers with `-g` options. This option returns the url `name`, `verb`, `path` and `reqs` field that match the pattern:
````
routers should be routes. I'd remove options after `-g`.

You make it sound like we return only the part of the route that matches - the name or verb etc. - instead of the whole route. I'm also unsure if we need to explain `grep`-ing. I think it makes the documentation clearer if we assume the reader knows that.
So how about: "You can grep your routes with `-g`, which filters routes by `name`, `verb`, `path`, or `reqs`."?

davydovanton replied:
yep, i like it :+1:

kaspth commented on railties/lib/rails/tasks/routes.rake:
```diff
@@ -1,7 +1,19 @@
+    opts.on("-g", "--grep PATTERN", String) do |pattern|
+      options[:pattern] = pattern
+    end
+  end.parse!(ARGV.reject { |x| x == "routes" })
```
Maybe not here, but `ARGV.delete('routes')`?

davydovanton replied:
But what if I call some rake tasks? For example:
```
$ rake restart routes ...
```

kaspth commented on railties/lib/rails/tasks/routes.rake:
```diff
@@ -1,7 +1,19 @@
+  end.parse!(ARGV.reject { |x| x == "routes" })
+
+  puts inspector.format(ActionDispatch::Routing::ConsoleFormatter.new, options)
+  exit 0
```
Why do we need the exit here?

davydovanton replied:
`exit` at the end will make sure that the extra arguments won't be interpreted as Rake task.

kaspth replied:
All right, let's add a comment for that.

kaspth commented on railties/test/application/rake_test.rb:
```diff
@@ -127,6 +127,18 @@ def test_rake_routes_with_controller_environment
+        end
+      RUBY
+
+      output = Dir.chdir(app_path){ `rake routes -- -g show` }
```
Didn't we support `-g` without `--`?

davydovanton replied:
it's my mistake, sorry :(

kaspth commented:
Do we even still want the CONTROLLER env option? Maybe we should just deprecate it instead.

davydovanton commented:
@kaspth I would like to see option which returns all urls to specific model or controller. In my rails app `admin` routes take up most part of all routes. And therefore I will be happy to see routes which pertain only for `Post` model/controller for example.

kaspth commented:
Btw, your commit says `-p` and not `-g` :smile:

It could be good to filter out a namespace. It's rare I want to look at admin routes.


> I will be happy to see routes which pertain only for Post model/controller for example.

Sounds interesting, how would users use this?

davydovanton commented:
> Btw, your commit says -p and not -g :smile:

@kaspth sorry, I little get sick therefore I wasn't noticed this :sweat: 

> Sounds interesting, how would users use this?

If you asked about syntax I think about something like this: 
```
rake routes (--controller | -c) (post | Post | PostController)
```

Also I added comment for `exit` and fixed typo in commit message :smiley:

kaspth commented:
Hope you're feeling better! :smile:

Sure, that could work. Does that include namespaces? Say `-c post` also finds `admin/posts` routes.

davydovanton commented:
@kaspth yes, I'm feeling better, thank you so much! :blush:
It's interesting question, thanks. I think that it will be better if `-c` option work without namespaces because we search by controller name and `PostController` and `AdminPostController` are differents controllers.

davydovanton commented:
@kaspth how do you think I should add a `-c` key feature in this PR or in a separate?

kaspth commented:
Go for it, and let's see how it works out :smile:

davydovanton commented:
@kaspth done :smiley:

kaspth commented on actionpack/CHANGELOG.md:
```diff
@@ -1,3 +1,11 @@
+*   Add `-g` and `-c` options for `rake routes` task. This options returns the url `name`,
+    `verb`, `path` and `reqs` field that match the pattern or returns routers for specific
```
It would be nice to mention that `-g` is short for grep and `-c` for controller.

kaspth commented on actionpack/lib/action_dispatch/routing/inspector.rb:
```diff
@@ -86,12 +85,15 @@ def format(formatter, filter = nil)
           @routes
         end
+
+        routes.select { |r| /#{options[:pattern]}/ =~ r.defaults.values.join(' ') }
```
Usually I'm used to seeing the pattern on the right hand side, why are you doing it the other way around? :smile:

kaspth replied:
Also variables default to nil in Ruby, so we don't need the else clause here:

```ruby
def filter_routes(options)
  controller_routes = if options[:controller]
    filter = options[:controller].downcase.sub(/_?controller\z/, '')
    @routes.select { |r| r.defaults[:controller] == filter }
  end
  
  (controller_routes || @routes).select { |r| /#{options[:pattern]}/ =~ r.defaults.values.join(' ') }
end
```

kaspth commented on guides/source/routing.md:
```diff
@@ -1112,10 +1112,18 @@ For example, here's a small section of the `rake routes` output for a RESTful ro
 
-You may restrict the listing to the routes that map to a particular controller setting the `CONTROLLER` environment variable:
-
+You may restrict the listing to the routes that map to a particular controller using `-c` option or setting the `CONTROLLER` environment variable:
```
I think we should deprecate the environment variable.

kaspth replied:
@davydovanton please remove the mention of `CONTROLLER` here, and in the example below. If you find it anywhere else please :zap: it too :smile:

kaspth replied:
`-c` is rather obtuse, instead let's say: 

```
using the `--controller` option — or its `-c` shorthand:
```

Then we should show both `--controller` and `-c` in the examples.

davydovanton replied:
done

kaspth commented on guides/source/routing.md:
````diff
@@ -1112,10 +1112,18 @@ For example, here's a small section of the `rake routes` output for a RESTful ro
 ```bash
 $ CONTROLLER=users bin/rake routes
+$ bin/rake routes -c users
+$ bin/rake routes -c users_controller
````
Does the `CONTROLLER` env var support the `_controller` suffix? If no, I don't think we have to. Besides the option is short for "search for a controller" so I think it's to just support "-c users" and infer that as "-c users_controller".

kaspth commented on railties/lib/rails/tasks/routes.rake:
```diff
@@ -1,7 +1,24 @@
+    opts.on("-c", "--controller [PATTERN]", String) do |pattern|
+      options[:controller] = pattern
+    end
+    opts.on("-g", "--grep [PATTERN]", String) do |pattern|
```
Isn't `String` the default? Do we need to write it?

kaspth commented on railties/lib/rails/tasks/routes.rake:
```diff
@@ -1,7 +1,24 @@
+
+  puts inspector.format(ActionDispatch::Routing::ConsoleFormatter.new, options)
+
+  # exit at the end will make sure that the extra arguments won't be interpreted as Rake task.
```
I'd say: `exit 0 # ensure extra arguments aren't interpreted as Rake tasks.`

kaspth commented on railties/test/application/rake_test.rb:
```diff
@@ -127,6 +127,30 @@ def test_rake_routes_with_controller_environment
       assert_equal "Prefix Verb URI Pattern     Controller#Action\n  cart GET  /cart(.:format) cart#show\n", output
     end
 
+    def test_rake_routes_with_controller_environment
```
This name doesn't match what you're testing - and there's a test just above it, which is called the same thing :smile:

kaspth commented on railties/test/application/rake_test.rb:
```diff
@@ -127,6 +127,30 @@ def test_rake_routes_with_controller_environment
+      assert_equal "Prefix Verb URI Pattern     Controller#Action\n  cart GET  /cart(.:format) cart#show\n", output
+
+      output = Dir.chdir(app_path){ `rake routes -c cart_controller` }
+      assert_equal "Prefix Verb URI Pattern     Controller#Action\n  cart GET  /cart(.:format) cart#show\n", output
```
`CONTROLLER` only supports the first way to write these, so let's keep it at that.

We should also test `-g` matches on the other parts of route it claims :wink:

Oh, and let's use two separate tests (or more if appropriate) for the `-g` and `-c` option respectively.

kaspth commented:
@dhh hey, @davydovanton has been adding `rake routes -g` here and also a `-c` option which searches for a controller (instead of the CONTROLLER env var). Is this close to what you're looking for?

sgrif commented:
Closing due to inactivity. There's a lot of unaddressed feedback. Let me know if you're still interested in working on this and I'll reopen.

sgrif commented:
Actually I'll leave this open for the time being so it stays on the radar. @davydovanton Are you still interested in working on this?

kaspth commented:
@davydovanton very much looking forward to have this in Rails 5. Are you still up for making it happen? :smile:

davydovanton commented:
@kaspth yes, of course! I'll try to complete this on weekend! :blush:

davydovanton commented:
@kaspth sorry for delay :sweat: 
I fix all your comments and I think now we can merge this :tada:
What do you think?

kaspth commented on actionpack/CHANGELOG.md:
```diff
@@ -1,3 +1,12 @@
+*   Add `-g` and `-c` (shorts for _grep_ and _controller_ respectively) options
+    for `rake routes` task. This options returns the url `name`, `verb`, `path`
+    and `reqs` field that match the pattern or returns routers for specific
```
What does `returns routers` mean?

kaspth commented on actionpack/lib/action_dispatch/routing/inspector.rb:
```diff
@@ -82,11 +81,19 @@ def format(formatter, filter = nil)
-          @routes
+        end
+
+        (filtered_routes || @routes).select { |r| /#{options[:pattern]}/ =~ r.defaults.values.join(' ') }
```
`/#{options[:pattern]}/` could be expensive for many routes, can we extract it to a local variable?

kaspth commented on actionpack/lib/action_dispatch/routing/inspector.rb:
```diff
@@ -82,11 +81,19 @@ def format(formatter, filter = nil)
 
-      def filter_routes(filter)
-        if filter
+      def filter_routes(options)
```
How about expanding options, like:

```ruby
def filter_routes(filter:, controller:, pattern:)
  filter ||= controller.downcase.sub(/_?controller\z/, '') if controller

  filtered_routes = if filter
    @routes.select { |route| route.defaults[:controller] == filter }
  else
    @routes
  end

  pattern = /#{pattern}/
  filtered_routes.select { |r| r.defaults.each_value.join(' '.freeze) =~ pattern }
end
```

That would get rid of the `filter_routes` method that calls `filter`, which feels weird :smile:

Sorry for backing down on removing the else branch that sets `@routes`, I think it's better the original way now :heart:

davydovanton replied:
It's interesting idea but I don't agree with you. With my solution you have simple way for adding new search pattern (for example, in future, it will be great idea to add search on actions).

kaspth replied:
I'm sorry I forgot to go back and update this with the ideas I had for the option parsing later. I'll do that now :smile:

I think it would be better to have the option parser do:

```ruby
opts.on("-c", "--controller [PATTERN]") do |pattern|
  options[:filter] = pattern.downcase.sub(/_?controller\z/, '')
end
```

Then later we can just assign `filter = options[:filter]` (or just use a keyword argument and save the line).

That's still as flexible as your proposal, but we only have to deal with one option. What do you think? :smile:

(If you think about it, what does `controller` really mean in the context of `filter_routes`? After all the other options are `filter` and `pattern` — `controller` sticks out like a sore thumb :grin:)

davydovanton replied:
yep, I understood. Nice idea :+1: 
Fixed.

kaspth replied:
Sorry for keeping this in limbo, but I don't think we should have both `filter` and `pattern`. In fact I think there's a way to get rid of `pattern`.

How about having the option parser do:

```ruby
opts.on("-c", "--controller [PATTERN]") do |pattern|
  options[: defaults_filter] = { controller: pattern.downcase.sub(/_?controller\z/, '') }
end

opts.on("-g", "--grep [PATTERN]") do |pattern|
  options[:defaults_filter] = ActionDispatch::Routing::ConsoleFormatter.defaults_filter_for(/#{pattern}/)
end
```

Later:

```ruby
if defaults_filter
  @routes.select do |route|
    defaults_filter.any? { |default, filter| route.defaults[default] =~ filter }
  end
else
  @routes
end
```

`defaults_filter_for` would just populate a hash where keys are defaults for a route, and the value is the pattern.

Make sense? :smile:

davydovanton replied:
Interesting idea. If I understand you right you want to set hash with all route parts in `defaults_filter_for`.
``` ruby
def defaults_filter_for(pattern)
  {
    controller: /#{options[:pattern]}/
    name: /#{options[:pattern]}/
    verb: /#{options[:pattern]}/
  }
  end
```

But I see one problem. Now, when you call something like this:
```
$ bin/rake routes -c user -g show
```
you'll see only `user_show` route. 

But your solution means that we will have similar hashes for `-c` key and for `-g` key:
``` ruby
# for -c
{ controller: pattern.downcase.sub(/_?controller\z/, '') }

# for -g
{ controller: /#{options[:pattern]}/ }
```

Therefore when we call `bin/rake routes` again, we will see routes only for `user`
controller or routes that match with `show`. What do you think?

kaspth replied:
Yeah, exactly, assigning the pattern to each route default and then matching on that.

Sure, it was made with the assumption that people won't really use `-g` and `-c` together. I can't really see a case where that makes sense, do you have any? :smile:

I think we could influence the behavior by how we frame the options in the documentation. I.e. saying `-c` is for when you have a specific controller in mind, and `-g` is for a general search. That way people would be less likely to mash them together.

kaspth replied:
@davydovanton have you had a chance to look into my comment? :smile:

kaspth replied:
@davydovanton have you had time to dig in here? :grin:

kaspth commented on railties/lib/rails/tasks/routes.rake:
```diff
@@ -1,7 +1,28 @@
+  OptionParser.new do |opts|
+    opts.banner = "Usage: rake routes [options]"
+    opts.on("-c", "--controller [PATTERN]") do |pattern|
+      options[:controller] = pattern
```
Perhaps we should move `downcase.sub(/_?controller\z/, '')` to here, and just set `options[:pattern]`. That would simplify the code further down a bit.

kaspth commented on railties/lib/rails/tasks/routes.rake:
```diff
@@ -1,7 +1,28 @@
-  puts inspector.format(ActionDispatch::Routing::ConsoleFormatter.new, ENV['CONTROLLER'])
+
+  if ENV['CONTROLLER']
+    ActiveSupport::Deprecation.warn('`CONTROLLER` environment is deprecated and will be removed in Rails 5.1. Please use -g and -c keys instead.')
```
I feel we need to explain the controller ENV variable better. It would be great to give up an upgrade example too.

How about:

```ruby
# At the top:
require 'active_support/deprecation'
require 'active_support/core_ext/string/strip' # for strip_heredoc

# ...

ActiveSupport::Deprecation.warn <<-eow.strip_heredoc
  Passing `CONTROLLER` to `bin/rake routes` is deprecated and will be removed in Rails 5.1.
  Please use `bin/rake routes --controller #{ENV['CONTROLLER']} instead.
eow
```

davydovanton replied:
looks cool!

kaspth commented on actionpack/CHANGELOG.md:
```diff
@@ -1,3 +1,12 @@
+
+    See #18902.
+
+    *Anton Davydov*
```
You're also deprecating `CONTROLLER` please mention that, and how to replace them. Perhaps you can find some inspiration in: https://github.com/rails/rails/pull/22173/files#diff-1846d4882fa63eb5b1271541995cbfdcR1 😁

kaspth commented:
@davydovanton thanks so much for getting back to it! I did find some more things which I think could be better, let me know what you think :smile:

davydovanton commented:
@kaspth I fixed all this and now code looks much better, thanks! :purple_heart:

kaspth commented on railties/test/application/rake_test.rb:
```diff
@@ -126,6 +126,36 @@ def test_rake_routes_with_controller_environment
       assert_equal "Prefix Verb URI Pattern     Controller#Action\n  cart GET  /cart(.:format) cart#show\n", output
     end
 
+    def test_rake_routes_with_global_search_key
```
Can you add `assert_deprecated` to the `ENV['CONTROLLER']` test just above this test?

kaspth commented on railties/test/application/rake_test.rb:
```diff
@@ -126,6 +126,36 @@ def test_rake_routes_with_controller_environment
+      output = Dir.chdir(app_path){ `rake routes -c Cart` }
+      assert_equal "Prefix Verb URI Pattern     Controller#Action\n  cart GET  /cart(.:format) cart#show\n", output
+
+      output = Dir.chdir(app_path){ `rake routes -c CartController` }
```
We only ever want to refer to `bin/rake` and not just `rake`. Please update your tests to match the existing ones :smile:

kaspth commented on guides/source/routing.md:
````diff
@@ -1134,10 +1134,17 @@ For example, here's a small section of the `rake routes` output for a RESTful ro
+$ bin/rake routes -c UsersController
+```
+
+You can grep your routes with `-g`, which filters routes by `name`, `verb`, `path`, or `reqs`.
````
What is `reqs` exactly? I assume it means requirements, but a quick Google search for `rails route requirements` found nothing in the current Rails guides — the only match was in 2.3.

I don't think we should mention it.

kaspth commented on actionpack/CHANGELOG.md:
```diff
@@ -1,3 +1,14 @@
+*   Add `-g` and `-c` (shorts for _grep_ and _controller_ respectively) options
+    for `rake routes` task. This options returns the url `name`, `verb`, `path`
```
We only ever want to refer to `bin/rake` and not just `rake`. Please update this.

kaspth commented on actionpack/CHANGELOG.md:
```diff
@@ -1,3 +1,14 @@
+*   Add `-g` and `-c` (shorts for _grep_ and _controller_ respectively) options
+    for `rake routes` task. This options returns the url `name`, `verb`, `path`
+    and `reqs` field that match the pattern or returns match for specific
```
As mentioned elsewhere I don't think we should mention `reqs`.

kaspth commented on actionpack/lib/action_dispatch/routing/inspector.rb:
```diff
@@ -82,12 +81,14 @@ def format(formatter, filter = nil)
+        filtered_routes = if filter
           @routes.select { |route| route.defaults[:controller] == filter }
-        else
-          @routes
```
Let's keep the else and avoid the git churn :smile:

davydovanton replied:
why you want this? :smile:

kaspth replied:
So we keep the git history intact :smile:

> Den 7. nov. 2015 kl. 18.53 skrev Anton Davydov <notifications@github.com>:
> 
> In actionpack/lib/action_dispatch/routing/inspector.rb <https://github.com/rails/rails/pull/20420#discussion_r44214912>:
> 
> >            @routes.select { |route| route.defaults[:controller] == filter }
> > -        else
> > -          @routes
> why you want this? 
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/rails/rails/pull/20420/files#r44214912>.
> 

--
Kasper

davydovanton commented:
@kaspth so, I removed `reqs` and canged `rake` to `bin/rake`. Also I added special test for check that `CONTROLLER` is deprecated.

kaspth commented on guides/source/routing.md:
````diff
@@ -1134,10 +1134,17 @@ For example, here's a small section of the `rake routes` output for a RESTful ro
+
+You can grep your routes with `-g`, which filters routes by `name`, `verb`, `path`, or `reqs`.
+```bash
+$ bin/rake routes -g users
````
I think it would make more sense to introduce the general searching first and then mention controller filtering. How about:


> You can search through your routes with the `--grep` option (`-g` for short). This outputs any routes that partially match the URL helper method name, the HTTP verb, or the URL path.
>
> ```bash
> $ bin/rake routes --grep new_comment
> $ bin/rake routes -g POST
> $ bin/rake routes -g admin
> ```
>
> If you only want to see the routes that map to a specific controller, there's the `--controller` option (`-c` for short).
>
> ```bash
> $ bin/rake routes --controller users
> $ bin/rake routes --controller admin/users
> $ bin/rake routes -c Comments
> $ bin/rake routes -c Articles::CommentsController
> ```

By the way, does the controller option support module namespaces like the last example?

kaspth commented on actionpack/lib/action_dispatch/routing.rb:
```diff
@@ -239,7 +239,8 @@ module ActionDispatch
   #
-  # Target specific controllers by prefixing the command with <tt>CONTROLLER=x</tt>.
+  # Target specific controllers by prefixing the command with <tt>--controller</tt> option
+  # - or its <tt>-c</tt> shorthand:.
```
There shouldn't be a `:` after shorthand here :smile:

davydovanton replied:
:sob:

kaspth commented on Gemfile.lock:
```diff
@@ -368,6 +368,3 @@ DEPENDENCIES
   w3c_validators
-
-BUNDLED WITH
-   1.10.6
```
Please don't remove this :smile:

kaspth commented on railties/test/application/rake_test.rb:
```diff
@@ -126,6 +126,47 @@ def test_rake_routes_with_controller_environment
       assert_equal "Prefix Verb URI Pattern     Controller#Action\n  cart GET  /cart(.:format) cart#show\n", output
     end
 
+    def test_rake_routes_with_controller_environment_is_deprecated
```
Please remove the `test_rake_routes_with_controller_environment` test :smile:

davydovanton replied:
Why? We still have this functional, and I think that we need remove this test after remove `CONTROLLER` env variable.

kaspth replied:
We shouldn't have two tests for the same thing when it's going away soon. Also I think it would actually be better to just bend the above test to add `assert_deprecated`.

davydovanton replied:
okay, I agree with you. Fixed.

kaspth commented on guides/source/routing.md:
````diff
@@ -1134,10 +1134,17 @@ For example, here's a small section of the `rake routes` output for a RESTful ro
+You can grep your routes with `-g`, which filters routes by `name`, `verb` or `path`.
 ```bash
-$ CONTROLLER=users bin/rake routes
+$ bin/rake routes -g users
````
I think it would make more sense to introduce the general searching first and then mention controller filtering. How about:

> You can search through your routes with the --grep option (-g for short). This outputs any routes that partially match the URL helper method name, the HTTP verb, or the URL path.
>
> ```bash
> $ bin/rake routes --grep new_comment
> $ bin/rake routes -g POST
> $ bin/rake routes -g admin
> ```
>
> If you only want to see the routes that map to a specific controller, there's the --controller option (-c for short).
>
> ```bash
> $ bin/rake routes --controller users
> $ bin/rake routes --controller admin/users
> $ bin/rake routes -c Comments
> $ bin/rake routes -c Articles::CommentsController
> ```

By the way, does the controller option support module namespaces like the last example?

davydovanton commented:
@kaspth I found [strange test](https://github.com/rails/rails/blob/f39ab9f3436a498f08c14578f5ea67591150bf27/actionpack/test/dispatch/routing/inspector_test.rb#L296-L311). In this test [`draw` test method](https://github.com/rails/rails/blob/f39ab9f3436a498f08c14578f5ea67591150bf27/actionpack/test/dispatch/routing/inspector_test.rb#L17-L21) tested. May be we can remove this test? What do you think?

kaspth commented on railties/test/application/rake_test.rb:
```diff
@@ -122,10 +122,43 @@ def test_rake_routes_with_controller_environment
 
       ENV['CONTROLLER'] = 'cart'
+
+      assert_deprecated { Dir.chdir(app_path){ `bin/rake routes` } }
```
Prefer initializing `output` outside the block, so we don't call the rake task twice:

```ruby
output = nil
assert_deprecated { output = Dir.chdir(app_path) { `bin/rake routes` } }
```

matthewd commented on railties/lib/rails/tasks/routes.rake:
```diff
@@ -1,7 +1,33 @@
+  if ENV['CONTROLLER']
+    ActiveSupport::Deprecation.warn <<-eow.strip_heredoc
+      Passing `CONTROLLER` to `bin/rake routes` is deprecated and will be removed in Rails 5.1.
+      Please use `bin/rake routes -c #{ENV['CONTROLLER']}` instead.
```
@kaspth FWIW, I'm really not a fan of including user-supplied values in deprecation messages. Even less so if we fail to quote them safely :neutral_face:

kaspth replied:
Ah right, hadn't thought about that. I'm fine with removing it :+1:

kaspth commented:
@davydovanton well, I do like that test checks every route for a resource, so we don't miss any. I'd just keep it.

kaspth commented:
@davydovanton had a chance to look into this further? :grin:

kaspth commented:
@davydovanton still up on making this happen for 5.0? We'd best get a move on then 😁

vipulnsward commented:
@kaspth @davydovanton going to work on a rebase for this to make it to 5.0
